version.go

```diff
@@ -142,40 +144,106 @@ func MustParseBinary(s string) Binary {
 
 // ParseBinary parses a binary version of the form "1.2.3-series-arch".
 func ParseBinary(s string) (Binary, error) {
-	m := binaryPat.FindStringSubmatch(s)
-	if m == nil {
+	groups := captureNamedGroups(s, binaryPat)
+	n := parseVersion(groups, true)
+	if n == nil {
 		return Binary{}, fmt.Errorf("invalid binary version %q", s)
 	}
-	var b Binary
-	b.Major = atoi(m[1])
-	b.Minor = atoi(m[2])
-	b.Tag = m[3]
-	b.Patch = atoi(m[4])
-	if m[5] != "" {
-		b.Build = atoi(m[5][1:])
-	}
-	b.Release = m[6]
-	b.Arch = m[7]
-	return b, nil
+
+	return Binary{
+		Number:  *n,
+		Release: groups["release"],
+		Arch:    groups["arch"],
+	}, nil
 }
 
-// Parse parses the version, which is of the form 1.2.3
-// giving the major, minor and release versions
-// respectively.
+// Parse a version in strict mode. The following version patterns are accepted:
+//  1.2.3       (major, minor, patch)
+//  1.2-tag3    (major, minor, patch, tag)
+//  1.2.3.4     (major, minor, patch, build)
+//  1.2-tag3.4  (major, minor, patch, build).
+//
+// The ParseNonStrict function can be used instead to parse a wider range of
+// version patterns (e.g. major only, major/minor etc.).
 func Parse(s string) (Number, error) {
-	m := numberPat.FindStringSubmatch(s)
-	if m == nil {
-		return Number{}, fmt.Errorf("invalid version %q", s)
+	groups := captureNamedGroups(s, numberPat)
+	if n := parseVersion(groups, true); n != nil {
+		return *n, nil
+	}
+	return Number{}, fmt.Errorf("invalid version %q", s)
+}
+
+// ParseNonStrict attempts to parse a version in non-strict mode. It supports
+// the same patterns as Parse with the addition of some extra patterns that
+// are not considered pure semantic version values.
+//
+// The following version patterns are accepted:
+//  1           (major)
+//  1.2         (major, minor)
+//  1.2.3       (major, minor, patch)
```

Comment:
I think that `1.2-tag` should be treated as valid (with a patch version of 0) by the **non-strict** parser. AFAICT from the strict parsing test cases, this pattern must be rejected in strict mode.

Thoughts?

Reply:
I would particularly reject this case. Said that, are we aware of any relevant reversion labelling following this convention?

Reply:
After applying the regex tweak suggested by @jameinel below, this is now being parsed in non-strict mode.